### ISSUE TITLE
Java master hangs on shutdown

### ISSUE THREAD

**Issue by antiquechrono** [dnp3/opendnp3#216]
MacOS 10.12.6
Java version "1.8.0_91"
Java(TM) SE Runtime Environment (build 1.8.0_91-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.91-b14, mixed mode)

I pulled and built the 2.0.x branch last night and was messing with the Java master-demo. When you type in quit it prints the following and never quits.

Channel state: CLOSED
Channel state: SHUTDOWN

I have not changed any of the code in the master-demo and am running it via 
mvn -pl example exec:java@master-demo

In my own code I have even tried calling System.exit(0) after master.shutdown() and it still hangs. If I remove the call to master.shutdown() then everything exits as expected.

I did a little profiling and it looks like it's getting stuck in com.automatak.dnp3.impl.ManagerImpl.shutdown_native_manager \[native] () 
and never exiting. I've never used JNI before so I'm not sure where this code lives to debug any further.

Also the outstation-demo exits correctly.

I was going to test this with the 2.2.0 tagged branch but it errors out compiling asiodnp3/ErrorCodes.cpp. If I had to guess this is probably due to the fact that Apple's Clang compiler is awful. I can report this in another issue if you want.

**Comment by jadamcrain** [comment]
Thanks for the report. I won't have time to test this out until week  12/11.  

That JNI method invokes the destructor for the C++ DNP3 Manger.  That calls terminates everything in an ordered way, and then calls join() for each thread in the thread pool.  It sounds like something is not cleaning up properly.

Do you get the same results if you run the C++ master demos on MaxOS?

**Comment by antiquechrono** [comment]
Sorry about the delay, I got really busy because of moving. When I try the C++ master and outstation demos they both exit cleanly.

**Comment by jadamcrain** [comment]
Version of Clang? I can try this version on Ubuntu and see if I can replicate.

**Comment by antiquechrono** [comment]
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 8.1.0 (clang-802.0.42)
Target: x86_64-apple-darwin16.7.0
Thread model: posix

Edit: If there's anything you would like me to try I would be happy to help.

**Comment by bbarcklay** [comment]
I'd like to report a similar issue using a [Python wrapper library](https://github.com/Kisensum/pydnp3) that was created with pybind11.  We've got a working Python outstation and master.  Everything works fine except for shutting down the outstation.  I've traced the call to DNP3Manager.Shutdown() and see the channel calling its ShutdownImpl and the MasterStack resource being freed but something is keeping io_service::run() blocked.  One theory is that shared pointer references in Python are keeping certain destructors from getting called but it looks like the call to DNP3Manager.Shutdown() should invoke all the shutdown code that is also called from destructors in classes like DNP3Channel and MasterStack.  Any ideas about where I might look or how I might determine what is causing the run() method to block? 

This occurs on both MacOS 10.13.3 and Linux/Ubuntu 16.

**Comment by jadamcrain** [comment]
It is probably caused by the same underlying issue in the C++, that is somehow surfaced by both bindings. I plan to review this ticket when I have time in mid-May.

**Comment by jadamcrain** [comment]
Update: I can replicate this on Linux.

**Comment by bbarcklay** [comment]
The hang on shutdown problem occurring in our [pydnp3](https://github.com/Kisensum/pydnp3) test code appears to be due to python references to opendnp3 objects, specifically the `channel` returned by `manager.AddTCPClient(...)` and the `master` returned by `channel.AddMaster(...)`, perhaps this is not surprising.  If we explicitly delete these references before calling `manager.Shutdown()`, everything works as expected.  I think the Python bindings increment a shared_ptr reference on these objects.  That may be interfering with a shutdown process that expects the destructors of these to objects to be executed.

**Comment by jadamcrain** [comment]
A user reported a similar issue in the C++ under the right conditions, which I have been able to resolve in this branch:

https://github.com/automatak/dnp3/tree/bug/deadlock_and_lifetimes

An infinite timer was being created that never exited, and this blocked the thread pool thread from exiting cleanly.

I just tested the master demo, and it also resolves this issue as well. This branch will get merged in 2.0.x for the upcoming 2.3.0-RC1 release.

**Comment by neilstephens** [comment]
Gday,

The merged fixes definitely improved the situation, but I can still produce a deadlock on cleanup using this example application built against 2.3.0 (700dc961d). And it crops up in the wild in my application tests for [opendatacon](https://github.com/neilstephens/opendatacon).

[main.cpp.txt](https://github.com/dnp3/opendnp3/files/3274196/main.cpp.txt)

The main thread gets stuck in ThreadPool::Shutdown(), blocked on joining the threads. The rest of the threads are still in ThreadPool::Run. So there must still be some asio task that's keeping everything from exiting.

I'll have a go at identifying the offending asio task by turning on asio debug output, but I haven't had much luck trying to interpret that output before. So maybe you guys can take another look?

Regards,
Neil.

**Comment by jadamcrain** [comment]
Thanks Neil. @emgre dived info the ASIO debug mode to fix some of the issues, so shouldn't be a problem finding the issue.

**Comment by neilstephens** [comment]
ASIO debug mode put me straight onto the source of the problem:

After it locked up, I could see that there was still a post, callback cycle happening every few seconds, so I was able to just put a breakpoint in asiopal::Executor::Post(), and sure enough, there is a SocketChannel re-transmit cycle that is stopping asio run from exiting.

![image](https://user-images.githubusercontent.com/9045862/59236868-53e53480-8c3b-11e9-84e3-46ab8e297b96.png)

I can see that asiodnp3::IOHandler::isShutdown bool is still false, so I'm assuming that is the problem. ie. Shutdown() on the SocketChannel hasn't been called in the shutdown sequence. How to fix it... that's where I get lost. Over to you :-)

Regards,
Neil.

**Comment by neilstephens** [comment]
I think I've given you a bum steer, sorry.

The example code is using two separate DNP3Managers (I've forgotten why I did that now - probably just ruling things in/out when I was troubleshooting the original problem).

So the re transmit cycle I was observing was on the other DNP3Manager. The DNP3Managers each have a separate asio::io_context, so they shouldn't impact each other shutting down... so I'm back to not knowing what's stopping asio run from returning.

Regards,
Neil.

**Comment by emgre** [comment]
@neilstephens On what platform/compiler are you experiencing this? I compiled your `main.cpp` on my Windows machine using MSVC 2019, both Debug and Release, and the program looks fine. It doesn't hang and use a constant amount of memory.

**Comment by neilstephens** [comment]
Ah, sorry. I should have mentioned that this is on Linux (it locks up on all Ubuntu and compiler versions I've tried).

That was a good reminder that @jadamcrain had to use Linux to recreate the original issue I raised. It also reminded me to go back and look at a long email chain we had going when we were first troubleshooting.

I'll copy you into the email @emgre later today when I'm not using my phone.

Reading back, I think Adam may have actually fixed this issue, but the other issues at the time were complicating things, so the changes from his troubleshooting branch were never merged.

It's Linux specific because I think it's a race between the shutdown sequence and the socket open/close sequence, which results in an epoll wait that can't be cancelled, whereas I think the equivalent in windows can be cancelled or has a timeout.

Cheers,
Neil

**Comment by emgre** [comment]
@neilstephens 
cc @jadamcrain 

I think I found the issue. With `ASIO_ENABLE_HANDLER_TRACKING` enabled, I was able to track down the handler that was preventing the thread pool to finish. It's in [TCPServer](https://github.com/dnp3/opendnp3/blob/master/cpp/libs/src/asiopal/TCPServer.cpp).

On Windows, if an acceptor already accepted a connection, putting the callback in ASIO handler queue, and you close the acceptor before being called back, the accepted socket becomes invalid. I added [this check](https://github.com/dnp3/opendnp3/blob/master/cpp/libs/src/asiopal/TCPServer.cpp#L112-L118) a [few commits ago](https://github.com/dnp3/opendnp3/commit/cb97034bbb1a494b0fac55b6105d942acbcbd908#diff-998cc72e85363c61abd997931be1f472) to prevent the fault.

On Linux, under the same circumstances, the socket is simply reported as valid. Therefore, the socket is accepted and the accept loop is continued even if the server was shutdown.

I'm working on a fix at the moment, I'll let you know when it's ready. It also affects the TLS server.

Thanks for the report!

**Comment by emgre** [comment]
You can try branch [`bugfix-tcpserver-shutdown`](https://github.com/dnp3/opendnp3/tree/bugfix-tcpserver-shutdown) and tell us if it fixes your issue.

**Comment by neilstephens** [comment]
Cool! I'll try out that branch in the next couple of hours. Sorry I didn't get around to forwarding those emails, but looks like you had enough info to go off anyway. Thanks.

**Comment by neilstephens** [comment]
Seems fixed to me. Nice work.

I'll have to look up some tutorials on using ASIO_ENABLE_HANDLER_TRACKING. It seems really powerful, and really neat how you can visualise what's going on with the handlerviz.pl script. I just have to figure out how to interpret the graphs and link the info back to the code.

Cheers,
Neil.

**Comment by emgre** [comment]
The fixes were merge in release [2.3.1](https://github.com/dnp3/opendnp3/releases/tag/2.3.1).
